src/System.Management.Automation/CoreCLR/CorePsPlatform.cs

```diff
@@ -204,6 +204,29 @@ internal static class CommonEnvVariableNames
 #endif
         }
 
+        /// <summary>
+        /// Get a temporary directory to use, needs to be unique to avoid collision
+        /// </summary>
+        public static string GetTemporaryDirectory()
+        {
+            string tempDir = "";
+            do
+            {
+                tempDir = Path.Combine(Path.GetTempPath(),System.Guid.NewGuid().ToString());
+            }
+            while (Directory.Exists(tempDir));
+
+            try
+            {
+                Directory.CreateDirectory(tempDir);
```

Comment:
There is still a race here - extremely unlikely, but another thread/process could create the same directory you're trying to create.